bin/propolis-server/src/lib/initializer.rs

```diff
@@ -994,6 +1001,121 @@ impl<'a> MachineInitializer<'a> {
         smb_tables.commit()
     }
 
+    fn generate_bootorder(&self) -> Result<Option<Entry>, Error> {
+        eprintln!(
+            "generating bootorder with order: {:?}",
+            self.boot_order.as_ref()
+        );
+        let Some(boot_names) = self.boot_order.as_ref() else {
+            return Ok(None);
+        };
+
+        let mut order = fwcfg::formats::BootOrder::new();
+
+        let parse_bdf =
+            |pci_path: &propolis_api_types::instance_spec::PciPath| {
+                let bdf: Result<pci::Bdf, Error> =
+                    pci_path.to_owned().try_into().map_err(|e| {
+                        Error::new(
+                            ErrorKind::InvalidInput,
+                            format!(
+                                "Couldn't get PCI BDF for {}: {}",
+                                pci_path, e
+                            ),
+                        )
+                    });
+
+                bdf
+            };
+
+        for boot_entry in boot_names.iter() {
+            if boot_entry.first_boot_only {
+                continue;
+            }
+            // names may refer to a storage device or network device.
+            //
+            // realistically we won't be booting from network devices, but leave that as a question
+            // for plumbing on the next layer up.
+
+            let storage_backend = |spec| {
+                let spec: &instance_spec::v0::StorageDeviceV0 = spec;
+                match spec {
+                    instance_spec::v0::StorageDeviceV0::VirtioDisk(disk) => {
+                        &disk.backend_name
+                    }
+                    instance_spec::v0::StorageDeviceV0::NvmeDisk(disk) => {
+                        &disk.backend_name
+                    }
+                }
+            };
+
+            let network_backend = |spec| {
+                let spec: &instance_spec::v0::NetworkDeviceV0 = spec;
+                let instance_spec::v0::NetworkDeviceV0::VirtioNic(vnic_spec) =
+                    spec;
+
+                &vnic_spec.backend_name
+            };
+
+            if let Some(device_spec) =
+                self.spec.devices.storage_devices.iter().find_map(
+                    |(_name, spec)| {
+                        if storage_backend(spec) == &boot_entry.name {
```

Comment:
oh, backend names are derived in Propolis! i re-found [`format!("{}-backend")`](https://github.com/oxidecomputer/propolis/blob/master/bin/propolis-server/src/lib/spec/api_request.rs#L76) that you'd fixed last week. at the very least this code is looking for backend names when it should be looking for device names.

FWIW i have a light preference towards using UUIDs as names here just to reduce the possibility of mishandling user input somewhere in, like, `sled-agent`. `1` we have to answer either way since we're turning one Nexus-provided name into two different identifiers (device vs backend), and `2`.. in a UUID-y world they probably _should_ be ephemeral-and-generated-on-start lest we have a category of device whose ID we _expect_ to match across all Propolis, versus the category we expect will always be unique globally.

i might have a stronger opinion about names here on the other end of this PR :) but until then i'm gonna put a pin in that thought.

Reply:
i have stronger opinions! as promised!

* if we use user-provided names anywhere, it feels much more intuitive for that to be a name visible in guests, rather than our own name for the Crucible volume backing a device by some name we invented. so i think we ought to flip `device_name` and `backend_name`.
* it'd be really nice to just get IDs from Nexus and not have to think about user-provided strings, lets figure out what needs doing for that.
* implicit component IDs probably should be defined by `propolis-server` unless we'd also have a runtime decoder ring of which ephemeral ID means what device for a given VM..

Reply:
(digested as https://github.com/oxidecomputer/propolis/issues/772)